### PR TITLE
(1422) Add a Type attribute to the Budget model

### PR DESCRIPTION
**Pull request by pezholio**
This adds a `type` attribute to the Budget model, allowing a user to specify if the funds come from the Newton or GCRF Fund. This can be extended to other fund types in the future. 

This PR (should have been two PRs TBH) has also removed the `period_start_date` and `period_end_date` columns from the UI in favour of a `financial_year` column. We then work out the Start and End dates based on the start and end of a given financial year. 

I've added a data migration to set the financial year for existing Budgets based on their start date. We'll need to remove the `period_start_date` and `period_end_date` columns from the DB once the data migrations have been run in staging and prod (I know in theory the data and DB migrations run in parallel, but I think it's safer to wait until everything is out in the wild)

### Before

image

### After

image

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -50,17 +50,16 @@
     t.boolean "call_present"
     t.date "call_open_date"
     t.date "call_close_date"
-    t.string "intended_beneficiaries", array: true
     t.string "roda_identifier_fragment"
     t.string "roda_identifier_compound"
+    t.string "intended_beneficiaries", array: true
```
Comment (leeky) [inline db/schema.rb:55]: Did you mean to commit these unrelated `db/schema.rb` changes?

File: spec/features/staff/users_can_create_a_budget_spec.rb [hunk 2]
```diff
@@ -150,10 +151,12 @@
     end
   end
 
-  def fill_in_and_submit_budget_form
+  def fill_in_and_submit_budget_form(financial_year: "#{Date.current.year}-#{Date.current.next_year.year}")
```
Comment (leeky) [inline spec/features/staff/users_can_create_a_budget_spec.rb:147]: `financial_year` doesn't get overridden anywhere, so is this needed?

File: app/models/financial_year.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+class FinancialYear
```
Comment (leeky) [inline app/models/financial_year.rb:1]: Looks like there might be some behaviour shared between this and `FinancialPeriod` (defined in `lib/financial_period.rb`).

File: app/models/financial_year.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      if [1, 2, 3].include?(date.month)
+        new(date.year - 1)
+      else
+        new(date.year)
+      end
+    end
+  end
+
+  def start_date
+    Date.parse("01-04-#{start_year}")
```
Comment (leeky) [inline app/models/financial_year.rb:24]: Would `Date.new(start_year, 4, 1)` make more sense here..?

File: spec/models/budget_spec.rb [hunk 5]
```diff
@@ -113,10 +115,16 @@
   context "when the period start and period end dates are invalid" do
     it "performs budget date validation" do
       budget = build(:budget, period_start_date: Date.today, period_end_date: Date.today + 366.days)
+  it "returns an instance of FinancialYear for the financial year" do
```
Comment (leeky) [inline spec/models/budget_spec.rb:79]: Might be worth using `travel_to` and locking this into an exact date and known values for the expectations?

File: db/data/20210127153525_set_financial_year_based_on_start_date.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+class SetFinancialYearBasedOnStartDate < ActiveRecord::Migration[6.0]
+  def up
+    Budget.all.each do |budget|
```
Comment (leeky) [inline db/data/20210127153525_set_financial_year_based_on_start_date.rb:3]: Probably worth doing `Budget.all.find_each` here instead.

Comment (leeky) [reply db/data/20210127153525_set_financial_year_based_on_start_date.rb:3]: Do we want to drop the `period_start_date` and `period_end_date` columns too?

**Review by leeky — APPROVED** [review]
Have made some comments, but :shipit: when ready!

File: db/schema.rb (hunk 1 above)
Comment (pezholio) [reply db/schema.rb:55]: Whoops, no!

File: app/models/financial_year.rb (hunk 4 above)
Comment (pezholio) [reply app/models/financial_year.rb:24]: Yeah, probably faster too!

File: app/models/financial_year.rb (hunk 3 above)
Comment (pezholio) [reply app/models/financial_year.rb:1]: Ah, good call. I've done a bit of refactoring to get rid of that module altogether, and introduce a new FinancialQuarter class. Would appreciate a look before merging if that's cool?

File: db/data/20210127153525_set_financial_year_based_on_start_date.rb (hunk 6 above)
Comment (pezholio) [reply db/data/20210127153525_set_financial_year_based_on_start_date.rb]: I deliberately skipped that. I don't 100% trust the Data Migrate gem, and would rather wait until the data migrations are done before dropping the column. Will make a card so we don't forget 👍

**Comment by pezholio** [comment]
@leeky Thanks for the review! ⭐ I know you've approved it, but I'd appreciate another quick look over https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/pull/881/commits/edd18bdeafa6f7ab04804475f89337c6e093828d, as I've introduced a bit of a refactor 👍

**Review by leeky — APPROVED** [review]
Recent refactors look good - have just added three small suggestions!

**Pull request merged**